demos/link-prediction/homogeneous-comparison-link-prediction.ipynb

```diff
@@ -0,0 +1,1515 @@
+{
```

Comment:
<p>@kjun9 found in https://github.com/stellargraph/stellargraph/pull/1614 that the average operator and even the Hadamard one probably was meaningless for these sort of algorithms, because the embeddings can end up rotated. What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1658/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Reply:
<p>From my observation, <code>average</code> operator indeed performs poorly compared with other operators. I agree that embedding can be rotated with the average operator. However, I think <code>Hadamard</code> is useful to measure node embedding difference. At least, <code>Hadamard</code> can tell us whether each dimensions of two embedding vectors share the same sign or not, i.e., the product of two positive or two negative figures is positive while the product of one positive and one negative figures is negative. For the embedding vectors generated by the <code>relu</code>  activation like GCN, Hadamard can indicate wether two embedding vectors have overlapped non-zero dimensions and which dimensions are overlapped. From my observation, <code>Hadamard</code> performs competitively compared with <code>operator_l1</code> and <code>operator_l2</code> and even performs better than them in some cases.</p>